src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -284,7 +284,29 @@ private void ChangeProjectProperty<T>(
 
                     if (!isFullyLoaded)
                     {
-                        TryReportCompilationThrownAway(_projectSystemProjectFactory.Workspace.CurrentSolution, Id);
+                        var currentSolution = _projectSystemProjectFactory.Workspace.CurrentSolution;
+                        var reportCompilationThrownAway = true;
+
+                        // Preprocessor directive only ParseOptions changes have special handling in DocumentState.UpdateParseOptionsAndSourceCodeKind.
+                        // We don't want to report telemetry for those changes or for the initial evaluation result.
+                        if (newValue is ParseOptions newParseOption)
+                        {
+                            if (oldValue is ParseOptions oldParseOptions)
+                            {
+                                var projectState = currentSolution.SolutionState.GetRequiredProjectState(Id);
+                                var syntaxTreeFactoryService = projectState.LanguageServices.GetRequiredService<ISyntaxTreeFactoryService>();
+
+                                if (syntaxTreeFactoryService.OptionsDifferOnlyByPreprocessorDirectives(oldParseOptions, newParseOption))
+                                    reportCompilationThrownAway = false;
+                            }
+                            else
+                            {
+                                reportCompilationThrownAway = false;
+                            }
+                        }
+
+                        if (reportCompilationThrownAway)
+                            TryReportCompilationThrownAway(currentSolution, Id);
                     }
                 }
             }
```

Comment:
If you could also make a paranoia change here: we should perhaps move all this sending of telemetry after the _projectPropertyModificationsInBatch.Add() line, so if this were to throw an exception, we're not corrupting state.

Reply:
Sounds good